rend3-routine/src/forward.rs

```diff
@@ -162,24 +162,60 @@ impl<M: Material> ForwardRoutine<M> {
         });
 
         let whole_frame_uniform_handle = builder.add_data(args.whole_frame_uniform_bg, NodeResourceUsage::Input);
-        let cull_handle = builder.add_data(args.culled, NodeResourceUsage::Input);
+        let cull_handle = builder.add_optional_data(args.culled, NodeResourceUsage::Input);
 
         builder.build(move |mut ctx| {
             let rpass = ctx.encoder_or_pass.take_rpass(rpass_handle);
             let whole_frame_uniform_bg = ctx.graph_data.get_data(ctx.temps, whole_frame_uniform_handle).unwrap();
-            let culled = match ctx.graph_data.get_data(ctx.temps, cull_handle) {
-                Some(c) => c,
-                None => return,
+
+            // If there exists data for the culling handle, that means we are running
+            // after culling has run. This means we are in secondary rendering.
+            let base_cull = match cull_handle {
+                Some(handle) => match ctx.graph_data.get_data(ctx.temps, handle) {
+                    Some(c) => Some(c),
+                    None => return,
+                },
+                None => None,
+            };
+            let mut culling_storage = ctx.data_core.graph_storage.get_mut(&self.culling_cache);
+            let secondary;
+            let culled = match base_cull {
+                Some(cull_set) => {
+                    culling_storage.insert(args.camera, cull_set.clone());
+                    secondary = true;
+                    cull_set
+                }
+                None => match culling_storage.get(&args.camera) {
+                    Some(cull_set) => {
+                        secondary = false;
+                        cull_set
+                    }
+                    None => return,
+                },
             };
+            let culled = ctx.temps.add(culled.clone());
+
+            // Counter-intuitively, we need to use the current object reference buffer here
```

Comment:
This is a super big gottcha, we need to figure out either naming or encapsulation to make this better.